### PR TITLE
Perform concretization of build-only deps separately

### PR DESCRIPTION
**Pull request by scheibelp**
This takes each dependency which has only the 'build' deptype and performs a distinct concretization on it. These build-only deps are no longer explicitly part of the dependency DAG. The dependency DAG may now include multiple instances of a given package, namely when it appears as a link dependency of the root and as a dependency of some package that is a build dependency. 

A common example is the cmake package: many packages may depend on cmake as a build dependency; cmake depends on zlib transitively. zlib is also a common dependency (transitively). With this PR, packages which depend on zlib via cmake and by linking via another package will build two instances of zlib.

It is not always possible to consider the concretization of a build-only dependency separately from other dependencies. At the moment this PR ignores this possibility and always performs a separate concretization.

TODOs:

- [x] add unit test for concretization as noted by @davydden 
- [x] ~~get rid of deptype 'nolink'~~ this was since handled by https://github.com/LLNL/spack/pull/2765
- [x] by default use same compiler for build deps when possible

See: 
* https://github.com/LLNL/spack/issues/839
* https://github.com/LLNL/spack/issues/2537

### CODE REVIEW

**Comment by davydden** [comment]
can you please add conretization test with 3 mockup packages:  
```
A -> B [build-only] -> C+variant
  \
    \ [link] -> C~variant
```
to make sure this feature won't get broken in future.

**Comment by adamjstewart** [comment]
> A common example is the cmake package: many packages may depend on cmake as a build dependency; cmake depends on zlib transitively. zlib is also a common dependency (transitively). With this PR, packages which depend on zlib via cmake and by linking via another package will build two instances of zlib.

I'm not sure I understand this example. Are you saying that if package A depends on zlib@1 and package B depends on cmake which depends on zlib@2, it will build two instances? Or are you saying it will always build two instances of zlib regardless of the version constraints? I think a more clear use case that has recently been a headache is trying to build Python 3 packages that have a link dependency on Python 3, but require GUI packages which have a build dependency on Python 2. In that case, the only way to build them is to build 2 different copies of Python, making this PR very useful.

**Comment by scheibelp** [comment]
> I think a more clear use case that has recently been a headache is trying to build Python 3 packages that have a link dependency on Python 3, but require GUI packages which have a build dependency on Python 2.

That will be another intended use case to satisfy with this - thanks for adding it!

> Are you saying that if package A depends on zlib@1 and package B depends on cmake which depends on zlib@2, it will build two instances? Or are you saying it will always build two instances of zlib regardless of the version constraints?

If the version of zlib needed by cmake and by package A were the same, it would not install the same zlib twice. However when concretizing zlib as part of cmake it will not account for any constraints on zlib placed by package A, which is to say if there was some zlib@1.5 that satisfied both cmake and package A, it would not be found in this PR.

**Comment by scheibelp** [comment]
> can you please add concretization test with 3 mockup packages

Added TODO - thanks @davydden

**Comment by citibeth** [comment]
> It is not always possible to consider the concretization of a build-only dependency separately from other dependencies. 

Can you expand on this please?

**Comment by scheibelp** [comment]
>>It is not always possible to consider the concretization of a build-only dependency separately from other dependencies.
>
>Can you expand on this please?

This occurs when a child of a build dependency conflicts with another dependency; what precisely constitutes a conflict is TBD - right now I'm just working with examples. For example:

X-[b]->Y-[r]->python2
X-[L, b]->python3

(note this particular case is not known to exist)

Here X is a python3 module which uses package Y to build. Y runs python2. In this case there is competition for PYTHONPATH.

**Comment by citibeth** [comment]
Wouldn't you need to concretize each build dependency separately?  In this
case, you would need to do two concretizations?:

1. X^python3 (included in the "main" DAG because it's also a link dep)
2. Y^python2  (build dep for X)

**Comment by becker33** [comment]
I think @scheibelp is worried about the case in which Y depends on python2 not as a build dep.

**Comment by scheibelp** [comment]
> Wouldn't you need to concretize each build dependency separately?

In this case even if you do, I think there is a question of how to properly set up PYTHONPATH when the time comes to build X - the separate concretizations would work fine but then there is an issue of properly referencing these dependencies from X.

X needs python3 and Y needs python2 to run, so for example I think it would be difficult to set up PYTHONPATH properly. Also setting up the reference to the 'python' executable (from the perspective of X) would have a conflict.

Perhaps most importantly I should say I don't think this will be a common issue; like I said for the example I put forth I don't know of any actual cases of it. I was going to put more effort in detecting these issues later on in the PR.

Another potential generic example could be:

X-[b]->Y-[L]->Z1
X-[L]->Z2

I say "could be" because I think RPATHs would generally avoid a conflict between Z1 and Z2, however if a system requires modifying LD_LIBRARY_PATH to reference libs (i.e. RPATH is not used) there would be a conflict. (NOTE: I'm not actually sure if that applies anywhere)

**Comment by citibeth** [comment]
> In this case even if you do, I think there is a question of how to properly set up PYTHONPATH when the time comes to build X

I see, thinks.  I suppose that how you set `$PYTHONPATH` is a fundamental problem, whether or not Spack is involved.  Such a package would be unbuildable, without some special Python Jujitsu in the package itself.

**Comment by scheibelp** [comment]
> I see, thinks. I suppose that how you set $PYTHONPATH is a fundamental problem, whether or not Spack is involved. Such a package would be unbuildable, without some special Python Jujitsu in the package itself.

Agreed - and to be clear I'm not focused on making such cases work for this PR, but perhaps I at least want to detect them.

**Comment by scheibelp** [comment]
This work is motivating a new dependency type: 'include'

This comes up for the libxcb dependency in python+tk. The following explicit dependencies exist

libxcb -[b, L]-> libxau -[b]-> xproto

There is also an implicit dependency (not in a package.py)

libxcb -[b]-> xproto

Build-only deps are presumed to essentially be tools that are run at build time and so can be built separately, for example building two separate instances of cmake is OK. xproto is a header library which should match between libxcb and libxau; even though it is only needed during build time it should not concretize separately. Therefore it should be an 'include' dependency.

In all the 4 deptypes would be:

* build: Add only to `PATH`, needs to be built for front-end. concretizes separately (unlike the other 3)
* include: Only needed for include at build time, not for linking
* link: Implies include, also sets RPATHs
* run: needed for running (e.g. also need to set PATH when used by build deps)

**Comment by adamjstewart** [comment]
We also need a `test` deptype. It would not be included in the hash, but would be necessary to build and include in the PATH when a user specifies `--run-tests`. See #1279

**Comment by scheibelp** [comment]
In it's current state this can build python+tk. In the next day or two it should also be useful to automatically create build dependencies on the frontend for systems with a frontend and backend architecture.

**Comment by scheibelp** [comment]
This is now able to automatically create build-only dependencies on the frontend for systems with a frontend and backend architecture. A test was added to check that the build-only dependencies are built with the same compiler as the top-level package if it is built for the frontend (or if there is no frontend/backend distinction).

1 test is currently failing because the cmake-client test package specifically references the 'cmake' build-only dependency. In general in this PR at the moment, one cannot place constraints on build-only dependencies (I anticipate being able to address this), for example the following will fail:

```
spack spec libxcb ^xcb-proto@1.12
```

I'll be back to address this on 1/3

**Comment by scheibelp** [comment]
Extra note: this PR also alters how deptypes determine environment configuration of a package for a build:

* Transitive link and include dependencies, as well as all direct build dependencies, are added to PKG_CONFIG_PATH
* Other than that, only build dependencies and their transitive run dependencies alter the environment. This includes adding their bin prefixes to PATH but also any other environment variables they would set

It is the opinion of this PR that only run dependencies (and build dependencies, which are run dependencies only needed at build time) should modify the environment

**Comment by scheibelp** [comment]
python modules which depended on py-setuptools to build are currently broken in this PR (for example py-cffi).

In my new handling of deptypes, the transitive dependencies of build dependencies are only included in environment setup if they are the run deptype. py-setuptools was not updated to include this deptype. The issue is not as simple as this though because once the run deptype is added for py-setuptools dependency on python, you end up with the following dependencies for a given python module X which depends on py-setuptools and python:

```
X -[b]-> py-setuptools -[b,r,L]-> python
X -[b,r,L]->python
```
	
This actually creates a potential conflict: build-only dependency of X on py-setuptools means that the python used for setuptools and the python used for X can be different; however both need to be present at the time that X is built (according to the deptypes), which creates a conflict.

The simplest way to handle this issue in the PR as of now is to add the "link" deptype to the X->py-setuptools dependency: this ensures that the same python is used as a direct dependency of X and as a transitive dependency via py-setuptools. I have modified python modules to get py-cffi to build to demonstrate this notion. I don't think this is a good way to proceed since it is a confusing way to use the "link" deptype. I'm currently pondering over a few different ways to proceed with this:

1. allow the concretizer to fall back on automatically choosing not to resolve a build-only dependency separately if it creates a conflict
2. allow the concretizer to specifically make an exception for py-setuptools: although this is less attractive than [1], approaches along the lines of [1] may end up looking like this if they compromise in terms of changing concretization logic.
3. add a build deptype that implies that the dependency should not be resolved separately

It will likely take at least a few days for me to figure out a path forward and start on that. I'll provide an update by the end of 4/6.

**Comment by luigi-calori** [comment]
Thanks a lot, @scheibelp I have verified I can build much of my stuff with your PR, as you mention, to test python packages, I' ve added link dependency to py-setuptools, and this workaround seem to work.
In my view this is not so bad if it means you have to use at runtime the same python you have used when installing with setuptools.
I've proceeed in testing and found another problem I have no idea of how to work around:
some of my needed deps return like:
spack spec libxext ---->
 Error: Separate concretization of xproto produced build-time conflict for PKG_CONFIG_PATH for libxext

**Comment by scheibelp** [comment]
>spack spec libxext ---->
Error: Separate concretization of xproto produced build-time conflict for PKG_CONFIG_PATH for libxext

Short answer: setting the deptype for xproto in libxext package to "include" should fix it.

That error message was added in this PR. It is part of logic that ensures that two separate instances of the same package (for example same package but different variants) are not concretized and then both added to PKG_CONFIG_PATH. All build dependencies (including build-only dependencies) and transitive link dependencies are added to PKG_CONFIG_PATH. Since the build-only dependencies are concretized separately this conflict becomes possible.

xproto is a header library and although it is only needed at build time, it should be concretized with other packages (the include deptype was added in this PR and is discussed earlier in this thread if you're interested in additional info). In this PR I set deptype to "include" for some packages in the libxext dependency tree but not libxext itself.

**Comment by scheibelp** [comment]
> 3. add a build deptype that implies that the dependency should not be resolved separately

Update: I pushed [3] at today's telecon and will be having a follow-up conversation on this tomorrow.

**Comment by samfux84** [comment]
"fix check for build dependencies in autotools packages" works for autoconf, automake and libtool, but for m4 I still get:

```
==> Error: RuntimeError: Cannot generate configure: missing dependencies ['m4']
/local/cluster/spack/lib/spack/spack/build_systems/autotools.py:208, in autoreconf:
     198      def autoreconf(self, spec, prefix):
     199          """Not needed usually, configure should be already there"""
     200          # If configure exists nothing needs to be done
     201          if os.path.exists(self.configure_abs_path):
     202              return
     203          # Else try to regenerate it
     204          autotools = ['m4', 'autoconf', 'automake', 'libtool']
     205          missing = [x for x in autotools if x not in spec.build_only_deps]
     206          if missing:
     207              msg = 'Cannot generate configure: missing dependencies {0}'
  >> 208              raise RuntimeError(msg.format(missing))
     209          tty.msg('Configure script not found: trying to generate it')
     210          tty.warn('*********************************************************')
     211          tty.warn('* If the default procedure fails, consider implementing *')
     212          tty.warn('*        a custom AUTORECONF phase in the package       *')
     213          tty.warn('*********************************************************')
     214          with working_dir(self.configure_directory):
     215              m = inspect.getmodule(self)
     216              # This part should be redundant in principle, but
     217              # won't hurt
     218              m.libtoolize()
     219              m.aclocal()
     220              # This line is what is needed most of the time
     221              # --install, --verbose, --force
     222              autoreconf_args = ['-ivf']
     223              if 'pkg-config' in spec:
     224                  autoreconf_args += [
     225                      '-I',
     226                      join_path(spec['pkg-config'].prefix, 'share', 'aclocal'),
     227                  ]
     228              autoreconf_args += self.autoreconf_extra_args
     229              m.autoreconf(*autoreconf_args)

See build log for details:
```

**Comment by scheibelp** [comment]
In this case I think the issue is that the bash-completion package does not declare m4 as a build dependency.

@alalazo is there a reason that the AutotoolsPackage doesn't declare build dependencies on m4, autoconf, automake, and libtool? Is the idea that most packages will have a "configure" script already so the ```AutotoolsPackage.autoreconf``` will exit early?

**Comment by citibeth** [comment]
Yes, most have a configure script.

**Comment by alalazo** [comment]
@scheibelp Yes, the idea is that most tarballs already come with a `configure` script and we don't want to add dependencies if not needed.

**Comment by adamjstewart** [comment]
> In this case I think the issue is that the bash-completion package does not declare m4 as a build dependency.

@alalazo and I had a conversation about this recently. Previously, it would check dependencies of dependencies. `m4` is a dependency of all of the other 3 packages, so it would be present whether or not it was a direct dependency. We'll have to check how many other packages relied on this fact. We can either add `m4` as a dependency to all, or remove the checks for `m4` being a direct dependency. I guess it depends on how deps of deps affect the environment these days.

**Comment by scheibelp** [comment]
> most have a configure script.

Got it - thanks @citibeth and @alalazo!

> We can either add m4 as a dependency to all, or remove the checks for m4 being a direct dependency. I guess it depends on how deps of deps affect the environment these days.

If a package needs to directly run commands provided by m4 during a build, then it should probably list it as a build dependency. Run dependencies of build dependencies should also be included in the environment (although they are not at the moment), so if m4 shows up as a run dependency of any of those packages and if any package that needs m4 would also need those packages then it wouldn't need to be listed.

For other folks who are reading through this: #3768 provides a spot to voice your opinion on deptypes

**Comment by svenevs** [comment]
Hey! I can confirm on a fresh spack installation on this branch that `python+tk` installs without problem.  However, this seems to have just deferred the problem elsewhere / this `python` cannot be used.  It seems directly related to this PR, but **please** let me know if you want me to open a separate issue instead so this thread doesn't get polluted.

For reference, I installed `python@3.6.0+tk`.  The resultant installation produced

```console
$ spack find -v
==> 23 installed packages.
-- linux-fedora23-x86_64 / gcc@5.3.1 ----------------------------
bzip2@1.0.6+shared  libpthread-stubs@0.3  libxcb@1.12           openssl@1.0.2k                   python@3.6.0+tk~ucs4  tcl@8.6.6           xcb-proto@1.12   xtrans@1.3.5
inputproto@2.3.2    libx11@1.6.3          libxdmcp@1.1.2        pkg-config@0.29.2+internal_glib  readline@7.0          tk@8.6.6            xextproto@7.3.0  zlib@1.2.11+pic+shared
kbproto@1.0.7       libxau@1.0.8          ncurses@6.0~symlinks  python@2.7.13~tk~ucs4            sqlite@3.18.0         util-macros@1.19.1  xproto@7.0.29
```

So we got `python~tk` and `python+tk` in there (**WOOOOOOT EFF YEAH!**).  The suspicious example, suspicious because _where the heck_ did `2.7:2.7.99,3.3:3.3.99` come from?

```console
$ spack spec py-ipython
Input spec
--------------------------------
py-ipython

Normalized
--------------------------------
py-ipython
    ^py-decorator
        ^python@2.7:2.7.99,3.3:3.3.999
            ^bzip2
            ^ncurses
            ^openssl
                ^zlib
            ^readline
            ^sqlite
    ^py-pathlib2
    ^py-pexpect
        ^py-ptyprocess
    ^py-pickleshare
    ^py-prompt-toolkit@1.0.4:1.999
        ^py-six@1.9.0:
        ^py-wcwidth
    ^py-pygments
    ^py-setuptools@18.5:
    ^py-simplegeneric@0.8:
    ^py-traitlets@4.2:
        ^py-enum34
        ^py-ipython-genutils

Concretized
--------------------------------
==> Error: Separate concretization of python produced build-time conflict
```

So if I try `spack spec py-decorator ^python /<my hash>` where `<my hash>` is either the `2.7.13~tk` or `3.6.0+tk`, it still gives the same error.  I can almost understand why the `+tk` might do this, but why the `~tk`?  P.S. I got to this point after bypassing the faulty `py-backports` dependency in #3937

**Comment by adamjstewart** [comment]
> where the heck did 2.7:2.7.99,3.3:3.3.99 come from?

I know what the problem is here. I'll submit a separate PR to fix this and you can try it out.

**Comment by scheibelp** [comment]
> So if I try ```spack spec py-decorator ^python /<my hash>``` where <my hash> is either the 2.7.13~tk or 3.6.0+tk, it still gives the same error.

I believe this is a concrete instance of the issue discussed above (https://github.com/LLNL/spack/pull/2548#issuecomment-291370345) where the following general dependency DAG cannot currently be concretized:

```
X -[b]-> py-setuptools -[b,r,L]-> python
X -[b,r,L]->python
```

There is more extensive discussion above but basically to get this PR in shape I have to update the Spack concretizer to merge the concretization of py-setuptools python with the concretization of the extended python. Lately I've placed a lower priority on this because @tgamblin is working on a concretizer update that is intended to be more general and handle this case (among others). A hack was discussed above where you can specify py-setuptools as a link dependency at which point the concretization should succeed.

**Comment by svenevs** [comment]
> A hack was discussed above where you can specify py-setuptools as a link dependency at which point the concretization should succeed.

Oh dear. After crawling through _every_ dependency and adding `link` to `py-setuptools` whenever found, concretization at long last completes (thank god for the `spack edit` command)!  Thanks, I misunderstood your [update](https://github.com/LLNL/spack/pull/2548#issuecomment-292326337) and thought that this was not necessary.

In all honesty, after a read-through I thought things were basically almost ready, which is why I posted in the first place -- to indicate that something was still broken.  Adam has told me of the things that have to take place for the new concretizer, and I fully appreciate how difficult this is!  As it stands, I'll just keep hacking things until they work.

I don't know how hard this would be, but one suggestion would be to (when the error occurs), explicitly state what the conflict(s) were.  It would be much more helpful in understanding what is breaking for the user :wink:

**Comment by adamjstewart** [comment]
> I don't know how hard this would be, but one suggestion would be to (when the error occurs), explicitly state what the conflict(s) were. It would be much more helpful in understanding what is breaking for the user 😉

This +1000. Although I imagine it's not trivial to do.

**Comment by svenevs** [comment]
Hey just wanted to put on your radar that `provides` may not be working as expected either.  I don't know the mechanics, but assume it's related to this stuff.  e.g. `py-matplotlib` will depend on `pil`, which should either be `py-pil` if python < 3.x, or `py-pillow` if >3.x.  It always grabs `py-pil` though (e.g. with my `python@3.6.0`).

**Comment by luigi-calori** [comment]
Uh.. again a conflict that prevent merging.
I currently rely on this PR for building python+tk
@tgamblin Is this PR going to be merged or there other different methods to address the problem?

File: lib/spack/spack/spec.py [hunk 1]
```diff
@@ -197,7 +198,7 @@
 _any_version = VersionList([':'])
 
 # Special types of dependencies.
-alldeps = ('build', 'link', 'run')
+alldeps = ('build', 'link', 'run', 'include')
 norun   = ('link', 'build')
```
Comment (adamjstewart) [inline lib/spack/spack/spec.py]: I thought we had gotten rid of these? @citibeth

**Comment by adamjstewart** [comment]
I've never seen `py-setuptools` as a link dependency before. Is this correct?

**Comment by scheibelp** [comment]
> I've never seen py-setuptools as a link dependency before. Is this correct?

Short answer: no, and that's why this is WIP

See https://github.com/LLNL/spack/pull/2548#issuecomment-296329214

**Comment by adamjstewart** [comment]
Ah yes, I've noticed this. When I write:
```python
extends('python')
depends_on('python', type=('build', 'run'))
```
it looks like it becomes `('build', 'link', 'run')`. If `extends` could support `type` and specific version specs, we would no longer need to combine them.

**Comment by luigi-calori** [comment]
Thanks a lot, I really need to build python+tk, with also other extensions, so this help.
I'm also ready with a PR on Paramiko.
Any suggestion how best practice to depend?

**Comment by scheibelp** [comment]
> I'm also ready with a PR on Paramiko. Any suggestion how best practice to depend?

To be clear, we're talking about deptypes (vs. depending on this PR)? 

#3768 suggests generally how I'd like things to depend on each other and what each deptype means. When I search for Paramiko in this repo I don't find results. Very briefly skimming http://www.paramiko.org/ suggests it is a python lib that perhaps links to python (or at least it builds c libs) so I'd guess the appropriate deptypes are build+link+run or potentially just build+run if it doesn't need to link to the python .so specifically. I believe build+run is the default for packages which extend PythonPackage

**Comment by adamjstewart** [comment]
You can submit your Paramiko PR, but don't include any of the changes from this PR. It's fine to merge as is, even if `python+tk` is still sketchy.

@tgamblin @scheibelp We mentioned this briefly in #4063, but is it worthwhile to comment out the Python dependencies in the X11 libs that cause problems with `python+tk`? We can add them back once this PR is merged. Until then, `python+tk` isn't really possible.

**Comment by scheibelp** [comment]
> You can submit your Paramiko PR, but don't include any of the changes from this PR. It's fine to merge as is, even if python+tk is still sketchy.

It looks like Paramiko itself doesn't require gui stuff so that makes sense

> @tgamblin @scheibelp We mentioned this briefly in #4063, but is it worthwhile to comment out the Python dependencies in the X11 libs that cause problems with python+tk? We can add them back once this PR is merged. Until then, python+tk isn't really possible.

Dependency on python is through ```xcb-proto``` which is a python module that is used to build stuff. Are you suggesting:

* Remove dependency on xcb-proto: I don't know if system python typically comes with this, if it doesn't then there could be an issue
* Remove xcb-proto dependency on python: this one would also be problematic I think because Spack would have to know how to set up the module
* Build with ~tk and maybe the system already has libx11 support (+tk doesn't change anything about the python compilation other than RPATHing in the spack-built x11 libs)

Overall I should say I'm certainly ok with removing dependencies and trusting things to the system in the meantime, and then fixing it when this support is more mature. Although it may not work either way.

**Comment by adamjstewart** [comment]
> Remove xcb-proto dependency on python: this one would also be problematic I think because Spack would have to know how to set up the module

This is the one I'm suggesting. `libxcb` also has a build dependency on Python 2 that we would probably have to comment out. The only downside is that the build may no longer work on systems where `python` is Python 3 (Gentoo). But I think it's worth it to get `python+tk` working in the short term.

**Comment by svenevs** [comment]
Hey, there might be a problem not worth actually solving, but worth explaining.  Currently, mixed python is going to give you an error:

```
Concretized
--------------------------------
==> Error: Separate concretization of python produced build-time conflict
```

This is generally probably what you want.  However, suppose I want to build `opencv+vtk` or `+qt`.  Both `vtk` and `qt` (in my experience), _cannot_ be built with `python3`.  More specifically, `vtk` has some hardcoded `py27` library trying to be built.

This now means that I am unable to build `opencv+python ^python@3.6.1`.  Opencv is kind of a special snowflake, so I'm not sure how many other packages will be affected by this.  The only workaround I can think of is currently you're gathering all of my build dependencies.  If something is _already_ built, this is not necessary.  E.g. since `opencv+qt` depends on `qt`, if I the user give you a specific hash for `qt` that is already built, don't inherit the `qt` build dependencies as well?

**Comment by scheibelp** [comment]
>  I am unable to build ```opencv+python ^python@3.6.1```

@svenevs 

It turns out this is another instance of https://github.com/LLNL/spack/pull/2548#issuecomment-296329214 that occurs specifically with py-numpy. Unfortunately since my error message is terrible it is hard for anyone to tell.

If I edit the py-setuptools dependency in py-numpy to be ```depends_on('py-setuptools', type=('build', 'link'))``` then ```spack spec -t opencv+python ^python@3.6.1``` succeeds.

At the moment I'm prioritizing fixing this issue over improving the error message (although both are essential) and moreover progress on this PR altogether has been slow on this because I've been prioritizing other work.

**Comment by svenevs** [comment]
> At the moment I'm prioritizing fixing this issue over improving the error message (although both are essential) and moreover progress on this PR altogether has been slow on this because I've been prioritizing other work.

Claro.  I will stop including additional information, I thought this was different.  I'm definitely not trying to rush anything...just trying to give user-testing!

**Comment by scheibelp** [comment]
> Claro. I will stop including additional information, I thought this was different. I'm definitely not trying to rush anything...just trying to give user-testing!

Whoops sorry I didn't mean to imply I don't want to know about issues, just that development is going slow on this. I really appreciate the testing folks have done on this! I definitely don't mind checking into problems people have.

I should also mention: part of the reason this has been prioritized lower is that @tgamblin may be able to incorporate support for this into his more comprehensive update to the concretization logic. That being said, issues that people want to report here will likely be useful as use cases for his work.

**Comment by svenevs** [comment]
Oh yeah no I meant to say with respect to the python stuff because it sounds like they all generally stem from the same thing!  I don't want to keep flooding this issue with things that are all the same cause, so if I have python specific issues I'm just gonna hack around them.

**Comment by syntion** [comment]
While working on #4319 I noticed, that `spack fetch -D` doesn't fetch the packages required for building.

**Comment by healther** [comment]
I just opened #6589 because of the unrestricted `py-setuptools` dependency in `py-flake8`, but add this means that we can no longer build our meta package (including `py-ipython` and `py-flake8`). 
Work around right now will probably be: exclude `py-ipython` from metapackage and manually add to the produced module/filesystemview. It would be nice if there would be an easier way.

**Comment by becker33** [comment]
Closing; superseded by #4595

**Pull request closed without merging**